images/virtualization-artifact/pkg/controller/vmop/snapshot/internal/step/validate_step.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2025 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package step
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/common/object"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/conditions"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service/restorer"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmop/snapshot/internal/common"
+	"github.com/deckhouse/virtualization-controller/pkg/eventrecord"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vmopcondition"
+)
+
+type ValidateStep struct {
+	client   client.Client
+	recorder eventrecord.EventRecorderLogger
+	cb       *conditions.ConditionBuilder
+}
+
+func NewValidateStep(
+	client client.Client,
+	recorder eventrecord.EventRecorderLogger,
+	cb *conditions.ConditionBuilder,
+) *ValidateStep {
+	return &ValidateStep{
+		client:   client,
+		recorder: recorder,
+		cb:       cb,
+	}
+}
+
+func (s ValidateStep) Take(ctx context.Context, vmop *v1alpha2.VirtualMachineOperation) (*reconcile.Result, error) {
+	if conditions.HasCondition(s.cb.GetType(), vmop.Status.Conditions) && s.cb.Condition().Status == metav1.ConditionTrue {
+		return nil, nil
+	}
+
+	if len(vmop.Status.Resources) > 0 {
+		return nil, nil
+	}
+
+	vmSnapshotKey := types.NamespacedName{Namespace: vmop.Namespace, Name: vmop.Spec.Restore.VirtualMachineSnapshotName}
+	vmSnapshot, err := object.FetchObject(ctx, vmSnapshotKey, s.client, &v1alpha2.VirtualMachineSnapshot{})
+	if err != nil {
+		common.SetPhaseConditionToFailed(s.cb, &vmop.Status.Phase, err)
+		return &reconcile.Result{}, err
+	}
+
+	if vmSnapshot.Status.VirtualMachineSnapshotSecretName == "" {
+		err := fmt.Errorf("snapshot secret name is empty")
+		common.SetPhaseConditionToFailed(s.cb, &vmop.Status.Phase, err)
+		return &reconcile.Result{}, err
+	}
```

Comment:
Is it possible to check for snapshot readiness and not fail if snapshot is in progress?